### PR TITLE
C#: Follow line directives when getting element location

### PR DESCRIPTION
**Pull request by tamasvajk**
~~This PR builds on https://github.com/github/codeql/pull/4976.~~

The main goal of this PR is to use locations of QL entities in non-generated files when possible. For example `.cshtml` files are compiled into `.cshtml.g.cs` files, and issues were reported in these generated files as these are the ones passed to csc. With this PR, locations are following `#line` directives. In case of aspnetcore, `.cshtml.g.cs` files point to the original `.cshtml` files.

This PR:
* improves `#line` directive extraction by including the referenced file in the source archive, and referencing the file in the directive entity,
* includes the same change for `#pragma checksum`,
* `Element::getLocation` was improved to take into account `#line` directives.

Missing
- [x] stats https://jenkins.internal.semmle.com/job/Stats/job/CSharp/33/
- [x] db upgrade folder
- [x] diff job ~~https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/817/~~  ~~https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/825/~~ https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/827/
- [x] change notes

### CODE REVIEW

File: csharp/ql/src/semmle/code/csharp/ExprOrStmtParent.qll [hunk 1]
```diff
@@ -57,7 +57,14 @@ cached
 private module Cached {
   cached
   Location bestLocation(Element e) {
-    result = e.getALocation().(SourceLocation)
+    result = e.getALocation().(SourceLocation) and
+    (
+      not exists(e.getALocation().(SourceLocation).getMappedLocation()) or
+      e instanceof LineDirective
```
[the remark this answers is not in the document]
Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/ExprOrStmtParent.qll:63]: Wouldn't it be better if `getMappedLocation` didn't return anything for `#line default`?

**Review by hvitved — COMMENTED** [review]
Looks great; only one question.

File: csharp/ql/src/semmle/code/csharp/ExprOrStmtParent.qll (hunk 1 above)
[the remark this answers is not in the document]
Comment (tamasvajk) [reply csharp/ql/src/semmle/code/csharp/ExprOrStmtParent.qll:63]: Yep, that would probably be a bit nicer. Would I do it like this?
```
  Location getMappedLocation() {
    locations_mapped(this, result) and
    not exists(LineDirective l | l.getALocation() = this)
  }
```

[the remark this answers is not in the document]
Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/ExprOrStmtParent.qll:63]: Yes, that looks correct.

**Review by hvitved — APPROVED** [review]

**Comment by tamasvajk** [comment]
@hvitved The latest differences job finished. It seems okay. Now issues show up in non-cs files, like this missing ternary opportunity here:
https://github.com/mono/mono/blob/69dbe53ae83148f4b65e0fa0bcf100b381c1345a/mcs/ilasm/parser/ILParser.jay#L863

There are three new extraction errors, all relating to reading a file. I pushed an extra commit to fix the logged message.

**Comment by tamasvajk** [comment]
I've rebased this PR and pushed a fix for the failing test.

**Review by hvitved — APPROVED** [review]

**Pull request merged**